### PR TITLE
CartesianTensor and method for conversion to IrrepTensor

### PR DESCRIPTION
**Pull request by blondegeek**

### CODE REVIEW

File: e3nn/tensor/cartesian_tensor.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+class CartesianTensor():
+    def __init__(self, tensor, formula=None):
+        self.N = len(tensor.shape)
+        if self.N > 13:
+            raise ValueError("CartesianTensor only supports up to 13 indices.")
+        assert tuple(tensor.shape) == tuple(3 for i in range(self.N))
+        self.base_indices = "abcdefghijklmnopqrstuvxyz"
+        if formula is None:
+            formula = self.base_indices[:self.N]
```
Comment (mariogeiger) [inline e3nn/tensor/cartesian_tensor.py:14]:
```suggestion
        if formula is None:
            formula = "abcdefghijklmnopqrstuvxyz"[:self.N]
```

File: e3nn/tensor/cartesian_tensor.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+from e3nn.tensor.irrep_tensor import IrrepTensor
+from e3nn import rs, o3
+
+
+class CartesianTensor():
+    def __init__(self, tensor, formula=None):
+        self.N = len(tensor.shape)
+        if self.N > 13:
+            raise ValueError("CartesianTensor only supports up to 13 indices.")
+        assert tuple(tensor.shape) == tuple(3 for i in range(self.N))
```
Comment (mariogeiger) [inline e3nn/tensor/cartesian_tensor.py:11]: I would add a message for the user :)

File: e3nn/tensor/cartesian_tensor.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
+import torch
+from e3nn.tensor.irrep_tensor import IrrepTensor
+from e3nn import rs, o3
+
+
+class CartesianTensor():
+    def __init__(self, tensor, formula=None):
+        self.N = len(tensor.shape)
+        if self.N > 13:
+            raise ValueError("CartesianTensor only supports up to 13 indices.")
```
Comment (mariogeiger) [inline e3nn/tensor/cartesian_tensor.py:10]: I would remove this but check below that len(formula) is equal to tensor.dim()

File: e3nn/tensor/cartesian_tensor.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+    def to_irrep_tensor(self):
+        basis_change = o3.xyz_vector_basis_to_spherical_basis()
+        new_indices = self.base_indices[self.N:2 * self.N]
+        old_indices = self.base_indices[:self.N]
+        rotation_pairs = "".join([a + b + "," for a, b in zip(new_indices, old_indices)])
+        einsum_str = rotation_pairs + old_indices + "->" + new_indices
+        irrep_basis = torch.einsum(einsum_str, *[basis_change] * self.N, self.tensor)
+        Rs = [(1, 1)]  # vectors
+        Rs_out, Q = rs.reduce_tensor(self.formula, **{i: Rs for i in old_indices})
```
Comment (mariogeiger) [inline e3nn/tensor/cartesian_tensor.py:26]: nothing ensures you that the formula provided by the user has been written with the first letters of the alphabet

File: e3nn/tensor/cartesian_tensor.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+        self.formula = formula
+        self.tensor = tensor
+
+    def to_irrep_tensor(self):
+        basis_change = o3.xyz_vector_basis_to_spherical_basis()
+        new_indices = self.base_indices[self.N:2 * self.N]
+        old_indices = self.base_indices[:self.N]
+        rotation_pairs = "".join([a + b + "," for a, b in zip(new_indices, old_indices)])
+        einsum_str = rotation_pairs + old_indices + "->" + new_indices
+        irrep_basis = torch.einsum(einsum_str, *[basis_change] * self.N, self.tensor)
```
Comment (mariogeiger) [inline e3nn/tensor/cartesian_tensor.py:24]: I would rename `irrep_basis` into simply `tensor`

File: tests/tensor/cartesian_tensor_test.py [hunk 6]
```diff
@@ -0,0 +1,18 @@
+# pylint: disable=not-callable, no-member, invalid-name, line-too-long, wildcard-import, unused-wildcard-import, missing-docstring
```
Comment (mariogeiger) [inline tests/tensor/cartesian_tensor_test.py:1]: Add a test with a formula provided by the user

File: e3nn/tensor/cartesian_tensor.py (hunk 3 above)
Comment (blondegeek) [reply e3nn/tensor/cartesian_tensor.py:10]: ah indeed! of course. thanks!

File: e3nn/tensor/cartesian_tensor.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (blondegeek) [reply e3nn/tensor/cartesian_tensor.py:24]: sleek!

File: e3nn/tensor/cartesian_tensor.py (hunk 4 above)
Comment (blondegeek) [reply e3nn/tensor/cartesian_tensor.py:25]: ah yes indeed!

File: e3nn/tensor/cartesian_tensor.py (hunk 2 above)
Comment (blondegeek) [reply e3nn/tensor/cartesian_tensor.py]: done.

File: e3nn/tensor/cartesian_tensor.py (hunk 5 above)
Comment (blondegeek) [reply e3nn/tensor/cartesian_tensor.py]: done

File: tests/tensor/cartesian_tensor_test.py (hunk 6 above)
Comment (blondegeek) [reply tests/tensor/cartesian_tensor_test.py:1]: done
